core/src/main/java/org/elasticsearch/common/xcontent/AbstractObjectParser.java

```diff
@@ -144,7 +137,7 @@ public void declareRawObject(BiConsumer<Value, BytesReference> consumer, ParseFi
     }
     private static <T> List<T> parseArray(XContentParser parser, IOSupplier<T> supplier) throws IOException {
         List<T> list = new ArrayList<>();
-        if (parser.currentToken().isValue()) {
+        if (parser.currentToken().isValue() || parser.currentToken() == XContentParser.Token.START_OBJECT) {
```

Comment:
the parsing of arrays allowed either the value directly, or an array of values, except in the case that values were represented by an object, which I think was wrong?